lxd/api_vsock.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/db/cluster"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/version"
+)
+
+var vsockAPI10Cmd = APIEndpoint{
+	Get: APIEndpointAction{Handler: vsockAPI10Get, AllowUntrusted: true},
+}
+
+var vsockConfigCmd = APIEndpoint{
+	Path: "config",
+	Get:  APIEndpointAction{Handler: vsockConfigGet, AllowUntrusted: true},
+}
+
+var vsockMetaDataCmd = APIEndpoint{
+	Path: "meta-data",
+	Get:  APIEndpointAction{Handler: vsockMetaDataGet, AllowUntrusted: true},
+}
+
+var vsockDevicesCmd = APIEndpoint{
+	Path: "devices",
+	Get:  APIEndpointAction{Handler: vsockDevicesGet, AllowUntrusted: true},
+}
+
+func authenticateAgentCert(d *Daemon, r *http.Request) (bool, string, error) {
+	cacheUpdated := false
+	trusted := false
+	projectInstanceName := ""
+
+	if len(d.agentCerts) == 0 {
+		// Cache is empty -> load instance and populate cache
+		err := updateAgentCertCache(d, r)
+		if err != nil {
+			return false, "", err
+		}
+
+		cacheUpdated = true
+	}
+
+	for _, cert := range r.TLS.PeerCertificates {
+		trusted, projectInstanceName = util.CheckTrustState(*cert, d.agentCerts, nil, false)
+		if trusted {
+			break
+		}
+	}
+
+	if !trusted {
+		// If the cache has been updated, but the certificate is not trusted, return Forbidden.
+		if cacheUpdated {
+			return false, "", nil
+		}
+
+		// Refresh the certificate cache as it might be out of date.
+		// Then try again.
+		err := updateAgentCertCache(d, r)
+		if err != nil {
+			return false, "", err
+		}
+
+		for _, cert := range r.TLS.PeerCertificates {
+			trusted, projectInstanceName = util.CheckTrustState(*cert, d.agentCerts, nil, false)
+			if trusted {
+				break
+			}
+		}
+
+		if !trusted {
+			return false, "", nil
+		}
+	}
+
+	return true, projectInstanceName, nil
+}
+
+func updateAgentCertCache(d *Daemon, r *http.Request) error {
```

Comment:
I think initially we should do away with this function and not try to maintain a cache. That will save us from having to figure out exactly when to invalidate it and the instance lookup with the optimized DB function should be fast enough.